pkg/apis/pipeline/v1beta1/task_validation.go

```diff
@@ -128,109 +95,123 @@ func ValidateDeclaredWorkspaces(workspaces []WorkspaceDeclaration, steps []Step,
-	for _, v := range volumes {
+	for idx, v := range volumes {
 		if vols.Has(v.Name) {
-			return &apis.FieldError{
+			errs = errs.Also(&apis.FieldError{
 				Message: fmt.Sprintf("multiple volumes with same name %q", v.Name),
 				Paths:   []string{"name"},
-			}
+			}).ViaIndex(idx)
+		} else {
+			vols.Insert(v.Name)
 		}
-		vols.Insert(v.Name)
 	}
-	return nil
+	return errs
 }
 
-func validateSteps(steps []Step) *apis.FieldError {
+func validateSteps(steps []Step) (errs *apis.FieldError) {
 	// Task must not have duplicate step names.
 	names := sets.NewString()
 	for idx, s := range steps {
 		if s.Image == "" {
-			return apis.ErrMissingField("Image")
+			errs = errs.Also(apis.ErrMissingField("Image")).ViaIndex(idx)
 		}
 
 		if s.Script != "" {
 			if len(s.Command) > 0 {
-				return &apis.FieldError{
+				errs = errs.Also(&apis.FieldError{
 					Message: fmt.Sprintf("step %d script cannot be used with command", idx),
 					Paths:   []string{"script"},
-				}
+				}).ViaIndex(idx)
 			}
 		}
 
 		if s.Name != "" {
 			if names.Has(s.Name) {
-				return apis.ErrInvalidValue(s.Name, "name")
+				errs = errs.Also(apis.ErrInvalidValue(s.Name, "name")).ViaIndex(idx)
+			}
+			if e := validation.IsDNS1123Label(s.Name); len(e) > 0 {
+				errs = errs.Also(&apis.FieldError{
+					Message: fmt.Sprintf("invalid value %q", s.Name),
+					Paths:   []string{"name"},
+					Details: "Task step name must be a valid DNS Label, For more info refer to https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names",
+				}).ViaIndex(idx)
 			}
 			names.Insert(s.Name)
 		}
 
 		for _, vm := range s.VolumeMounts {
 			if strings.HasPrefix(vm.MountPath, "/tekton/") &&
 				!strings.HasPrefix(vm.MountPath, "/tekton/home") {
-				return &apis.FieldError{
+				errs = errs.Also(&apis.FieldError{
 					Message: fmt.Sprintf("step %d volumeMount cannot be mounted under /tekton/ (volumeMount %q mounted at %q)", idx, vm.Name, vm.MountPath),
 					Paths:   []string{"volumeMounts.mountPath"},
-				}
+				}).ViaIndex(idx)
 			}
 			if strings.HasPrefix(vm.Name, "tekton-internal-") {
-				return &apis.FieldError{
+				errs = errs.Also(&apis.FieldError{
 					Message: fmt.Sprintf(`step %d volumeMount name %q cannot start with "tekton-internal-"`, idx, vm.Name),
 					Paths:   []string{"volumeMounts.name"},
-				}
+				}).ViaIndex(idx)
 			}
 		}
 	}
-	return nil
+	return errs
 }
 
-func ValidateParameterTypes(params []ParamSpec) *apis.FieldError {
+func ValidateParameterTypes(params []ParamSpec) (errs *apis.FieldError) {
 	for _, p := range params {
-		// Ensure param has a valid type.
-		validType := false
-		for _, allowedType := range AllParamTypes {
-			if p.Type == allowedType {
-				validType = true
-			}
-		}
-		if !validType {
-			return apis.ErrInvalidValue(p.Type, fmt.Sprintf("taskspec.params.%s.type", p.Name))
+		errs = errs.Also(p.ValidateType())
+	}
+	return errs
+}
+
+func (p ParamSpec) ValidateType() *apis.FieldError {
```

Comment:
This should move on a different `_validation.go` file :wink:

Reply:
yup would be great to move such functions 🙏